content/articles/how-to-write-efficient-python-code/index.md

```diff
@@ -0,0 +1,242 @@
+title: How to write efficient Python code
+description: This tutorial provides the reader with detailed guid on how to use List Comprehensions, Lambda functions and Array functions to write efficiently as a python developer.
+author: samuel-torimiro
+date: 2021-05-06T00:00:00-09:00
+topics: [Python]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/how-to-write-efficient-python-code/hero.jpg
+    alt: Python example image
+---
+As a beginner programmer, writing efficient python code is not the main goal but rather to learn the basic syntax, the flow of the language, and how to think like a programmer. But things get interesting when you are trying to advance to an intermediate-level developer and become a true nerd, no pun intended.
+<!--more-->
+
+Writing efficient python code is one of the ways a beginner programmer can advance to become an intermediate programmer and appreciate code readability coupled with the productivity of a developer.
+
+This tutorial is focused on teaching you how to write efficient python code. By the end of this tutorial, you will be equipped with the tools needed to improve your productivity as a developer. Tools we would be using include `list comprehension`, `lambda`, etc.
+
+### Prerequisites
+To follow along with this tutorial, a basic understanding of python is required. A python interpreter should be installed in your local environment.
+
+### List Comprehension
+List comprehension is a technique of creating a list containing data with a single line of code. It's common in various programming languages including Python. 
+
+Let's go over the syntax of list comprehension;
+
+\**result*\* = [\**transform*\* \**iteration*\* \**filter*\*]
+
+The *result* would be the final list containing our data, the *transform* is the value of each data in our list, it keeps on changing as the value of the iterator changes. The *iteration* is a loop that helps populate our list with the required amount of data. Finally, the *filter* (optional) is used to filter out the data that we don't want.
+
+Before we take a look at an example of list comprehension, let's take a look at the traditional way of generating a list with data so we can compare the two ways.
+
+```python
+nums = []
+for x in range(10):
+    nums.append(x)
+
+print(nums)
+```
+
+**Output:**
+
+```bash
+[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+```
+It took us about 3 lines of code. Let's try to be more efficient using `list comprehension`.
+
+```python
+nums = [x for x in range(10)]
+
+print(nums)
+```
+
+**Output:**
+
+```bash
+[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+```
+
+Voilà, just one line of code to achieve the earlier result. It is way quicker to write once you understand the syntax and flow of `list comprehension`.
+
+Let's take a look at two more complex examples by adding *filters*.
+
+```python
+nums = [x for x in range(20) if x % 2 != 0]
+
+print(nums)
+```
+
+**Output:**
+
+```bash
+[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

Comment:
How did you do the technical review? @zolomohan 
The output is wrong. It should be:

```bash
[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]
```

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

Reply:
Oh shoot. Must have missed it. I ran all the snippets in an online compiler. I'll fix it and double check the other outputs too. Sorry for the inconvenience.

Reply:
Don't worry about it, I already fixed them. Will commit it soon

Reply:
thank you so much.